File: rust/onp-core/src/session.rs
//! Session lifecycle, cryptographic handshake state machine, and anti-replay protection.

use rand::rngs::OsRng;
use rand::RngCore;

use crate::constants::{
    flags, ENVELOPE_HEADER_SIZE, HANDSHAKE_NONCE_SIZE, LogicalOpcode, MAC_TAG_SIZE, NONCE_SIZE,
};
use crate::crypto::{
    decrypt_aead, derive_session_keys, encrypt_aead, KeyExchange,
};
use crate::errors::{OnpError, OnpResult};
use crate::framing::{construct_nonce, extract_sequence, EnvelopeHeader};
use crate::opcode::OpcodeTable;
use crate::packet::{ApplicationPacket, HandshakeAck, HandshakeSyn};

/// Role assumed by an endpoint in the ONP connection.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SessionRole {
    Client,
    Server,
}

/// Internal state machine of the session handshake.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SessionState {
    Uninitialized,
    SynSent,
    Established,
}

/// 64-packet sliding window filter to block replay attacks in O(1) time.
#[derive(Debug, Clone, Default)]
pub struct AntiReplayWindow {
    max_sequence: u64,
    bitmap: u64,
}

impl AntiReplayWindow {
    pub fn new() -> Self {
        Self {
            max_sequence: 0,
            bitmap: 0,
        }
    }

    /// Validates whether a sequence number is acceptable and updates window state.
    pub fn check_and_update(&mut self, sequence: u64) -> bool {
        if sequence == 0 {
            return false;
        }

        if sequence > self.max_sequence {
            let diff = sequence - self.max_sequence;
            if diff < 64 {
                self.bitmap = (self.bitmap << diff) | 1;
            } else {
                self.bitmap = 1;
            }
            self.max_sequence = sequence;
            true
        } else {
            let diff = self.max_sequence - sequence;
            if diff >= 64 {
                // Too old, fell outside sliding window
                false
            } else {
                let bit = 1u64 << diff;
                if (self.bitmap & bit) != 0 {
                    // Already received
                    false
                } else {
                    self.bitmap |= bit;
                    true
                }
            }
        }
    }
}

/// Complete cryptographically authenticated and polymorphic session manager.
pub struct OnpSession {
    pub role: SessionRole,
    pub state: SessionState,
    key_exchange: Option<KeyExchange>,
    local_nonce: [u8; HANDSHAKE_NONCE_SIZE],
    peer_nonce: [u8; HANDSHAKE_NONCE_SIZE],
    send_key: Option<[u8; 32]>,
    recv_key: Option<[u8; 32]>,
    send_sequence: u64,
    replay_window: AntiReplayWindow,
    opcode_table: Option<OpcodeTable>,
    epoch: u32,
}

impl OnpSession {
    /// Creates a new client session.
    pub fn new_client() -> Self {
        let mut local_nonce = [0u8; HANDSHAKE_NONCE_SIZE];
        OsRng.fill_bytes(&mut local_nonce);
        Self {
            role: SessionRole::Client,
            state: SessionState::Uninitialized,
            key_exchange: Some(KeyExchange::new()),
            local_nonce,
            peer_nonce: [0u8; HANDSHAKE_NONCE_SIZE],
            send_key: None,
            recv_key: None,
            send_sequence: 0,
            replay_window: AntiReplayWindow::new(),
            opcode_table: None,
            epoch: rand::random(),
        }
    }

    /// Creates a new server session.
    pub fn new_server() -> Self {
        let mut local_nonce = [0u8; HANDSHAKE_NONCE_SIZE];
        OsRng.fill_bytes(&mut local_nonce);
        Self {
            role: SessionRole::Server,
            state: SessionState::Uninitialized,
            key_exchange: Some(KeyExchange::new()),
            local_nonce,
            peer_nonce: [0u8; HANDSHAKE_NONCE_SIZE],
            send_key: None,
            recv_key: None,
            send_sequence: 0,
            replay_window: AntiReplayWindow::new(),
            opcode_table: None,
            epoch: rand::random(),
        }
    }

    /// Returns true if the session handshake is established and ready for encrypted traffic.
    pub fn is_established(&self) -> bool {
        self.state == SessionState::Established
    }

    /// CLIENT STEP 1: Generates the wire-encoded `HANDSHAKE_SYN` frame.
    pub fn create_handshake_syn(&mut self) -> OnpResult<Vec<u8>> {
        if self.role != SessionRole::Client {
            return Err(OnpError::HandshakeStateError {
                state: "ServerCannotInitiateSyn",
                flags: flags::HANDSHAKE_SYN,
            });
        }

        let kx = self.key_exchange.as_ref().ok_or_else(|| {
            OnpError::KeyExchangeFailed("Key exchange state missing".into())
        })?;

        let syn = HandshakeSyn {
            client_public_key: kx.public_key,
            client_nonce: self.local_nonce,
        };

        let payload = syn.encode();
        let header = EnvelopeHeader::new(flags::HANDSHAKE_SYN, payload.len() as u32);

        let mut frame = Vec::with_capacity(ENVELOPE_HEADER_SIZE + payload.len());
        frame.extend_from_slice(&header.encode());
        frame.extend_from_slice(&payload);

        self.state = SessionState::SynSent;
        Ok(frame)
    }

    /// SERVER STEP 2: Processes incoming `HANDSHAKE_SYN` and generates the response `HANDSHAKE_ACK` frame.
    pub fn process_handshake_syn(&mut self, syn_frame: &[u8]) -> OnpResult<Vec<u8>> {
        if self.role != SessionRole::Server {
            return Err(OnpError::HandshakeStateError {
                state: "ClientCannotProcessSyn",
                flags: flags::HANDSHAKE_SYN,
            });
        }

        let header = EnvelopeHeader::decode(syn_frame)?;
        if header.flags & flags::HANDSHAKE_SYN == 0 {
            return Err(OnpError::HandshakeStateError {
                state: "ExpectedHandshakeSynFlag",
                flags: header.flags,
            });
        }

        let syn_payload = &syn_frame[ENVELOPE_HEADER_SIZE..];
        let syn = HandshakeSyn::decode(syn_payload)?;
        self.peer_nonce = syn.client_nonce;

        let kx = self.key_exchange.take().ok_or_else(|| {
            OnpError::KeyExchangeFailed("Server key exchange already consumed".into())
        })?;
        let server_public_key = kx.public_key;

        // Perform Diffie-Hellman
        let shared_secret = kx.complete(&syn.client_public_key)?;

        // Derive Keys: ClientWriteKey, ServerWriteKey, OpcodeSeed
        let keys = derive_session_keys(&shared_secret, &self.peer_nonce, &self.local_nonce)?;

        // For server: send_key = server_write_key, recv_key = client_write_key
        self.send_key = Some(keys.server_write_key);
        self.recv_key = Some(keys.client_write_key);
        self.opcode_table = Some(OpcodeTable::from_seed(&keys.opcode_seed));
        self.state = SessionState::Established;

        let ack = HandshakeAck {
            server_public_key,
            server_nonce: self.local_nonce,
        };

        let ack_payload = ack.encode();
        let ack_header = EnvelopeHeader::new(flags::HANDSHAKE_ACK, ack_payload.len() as u32);

        let mut frame = Vec::with_capacity(ENVELOPE_HEADER_SIZE + ack_payload.len());
        frame.extend_from_slice(&ack_header.encode());
        frame.extend_from_slice(&ack_payload);

        Ok(frame)
    }

    /// CLIENT STEP 3: Processes incoming `HANDSHAKE_ACK` from server to finalize connection.
    pub fn process_handshake_ack(&mut self, ack_frame: &[u8]) -> OnpResult<()> {
        if self.role != SessionRole::Client || self.state != SessionState::SynSent {
            return Err(OnpError::HandshakeStateError {
                state: "ClientNotInSynSentState",
                flags: flags::HANDSHAKE_ACK,
            });
        }

        let header = EnvelopeHeader::decode(ack_frame)?;
        if header.flags & flags::HANDSHAKE_ACK == 0 {
            return Err(OnpError::HandshakeStateError {
                state: "ExpectedHandshakeAckFlag",
                flags: header.flags,
            });
        }

        let ack_payload = &ack_frame[ENVELOPE_HEADER_SIZE..];
        let ack = HandshakeAck::decode(ack_payload)?;
        self.peer_nonce = ack.server_nonce;

        let kx = self.key_exchange.take().ok_or_else(|| {
            OnpError::KeyExchangeFailed("Client key exchange already consumed".into())
        })?;

        // Perform Diffie-Hellman
        let shared_secret = kx.complete(&ack.server_public_key)?;

        // Derive Keys
        let keys = derive_session_keys(&shared_secret, &self.local_nonce, &self.peer_nonce)?;

        // For client: send_key = client_write_key, recv_key = server_write_key
        self.send_key = Some(keys.client_write_key);
        self.recv_key = Some(keys.server_write_key);
        self.opcode_table = Some(OpcodeTable::from_seed(&keys.opcode_seed));
        self.state = SessionState::Established;

        Ok(())
    }

    /// Encrypts an application command into a complete ONP wire frame.
    /// Maps the logical opcode to the polymorphic physical ID, generates a monotonic nonce,
    /// and encrypts using ChaCha20-Poly1305 with the envelope header as AAD.
    pub fn encrypt_packet(
        &mut self,
        opcode: LogicalOpcode,
        payload: &[u8],
    ) -> OnpResult<Vec<u8>> {
        if self.state != SessionState::Established {
            return Err(OnpError::HandshakeStateError {
                state: "SessionNotEstablished",
                flags: 0,
            });
        }

        let send_key = self.send_key.as_ref().unwrap();
        let opcode_table = self.opcode_table.as_mut().unwrap();

        // 1. Resolve polymorphic physical opcode
        let physical_opcode = opcode_table.to_physical(opcode);

        // 2. Prepare plaintext = [PhysicalOpcode (2B LE) || Payload]
        let mut plaintext = Vec::with_capacity(2 + payload.len());
        plaintext.extend_from_slice(&physical_opcode.to_le_bytes());
        plaintext.extend_from_slice(payload);

        // 3. Monotonic sequence counter
        self.send_sequence += 1;
        let nonce = construct_nonce(self.epoch, self.send_sequence);

        // 4. Header: Flags = ENCRYPTED, PayloadLen = NONCE_SIZE (12) + CIPHERTEXT + MAC_TAG_SIZE (16)
        let total_encrypted_len = (NONCE_SIZE + plaintext.len() + MAC_TAG_SIZE) as u32;
        let header = EnvelopeHeader::new(flags::ENCRYPTED, total_encrypted_len);
        let header_bytes = header.encode();

        // 5. Encrypt with AAD = Header
        let ciphertext_and_tag = encrypt_aead(send_key, &nonce, &header_bytes, &plaintext)?;

        // 6. Assemble frame: [Header (8B) || Nonce (12B) || Ciphertext+Tag]
        let mut frame = Vec::with_capacity(ENVELOPE_HEADER_SIZE + total_encrypted_len as usize);
        frame.extend_from_slice(&header_bytes);
        frame.extend_from_slice(&nonce);
        frame.extend_from_slice(&ciphertext_and_tag);

        Ok(frame)
    }

    /// Decrypts an incoming ONP wire frame, checks anti-replay window, verifies AEAD tag,
    /// and maps the polymorphic opcode back to the application logical opcode.
    pub fn decrypt_packet(&mut self, frame: &[u8]) -> OnpResult<ApplicationPacket> {
        if self.state != SessionState::Established {
            return Err(OnpError::HandshakeStateError {
                state: "SessionNotEstablished",
                flags: 0,
            });
        }

        let header = EnvelopeHeader::decode(frame)?;
        if header.flags & flags::ENCRYPTED == 0 {
            return Err(OnpError::HandshakeStateError {
                state: "ExpectedEncryptedFlag",
                flags: header.flags,
            });
        }

        let expected_frame_len = ENVELOPE_HEADER_SIZE + header.payload_len as usize;
        if frame.len() < expected_frame_len {
            return Err(OnpError::BufferTooShort {
                expected: expected_frame_len,
                actual: frame.len(),
            });
        }

        let encrypted_block = &frame[ENVELOPE_HEADER_SIZE..expected_frame_len];
        if encrypted_block.len() < NONCE_SIZE + MAC_TAG_SIZE + 2 {
            return Err(OnpError::BufferTooShort {
                expected: NONCE_SIZE + MAC_TAG_SIZE + 2,
                actual: encrypted_block.len(),
            });
        }

        // 1. Extract 12-byte Nonce and verify anti-replay sequence
        let mut nonce = [0u8; NONCE_SIZE];
        nonce.copy_from_slice(&encrypted_block[..NONCE_SIZE]);
        let sequence = extract_sequence(&nonce);

        if !self.replay_window.check_and_update(sequence) {
            return Err(OnpError::ReplayDetected { sequence });
        }

        // 2. Decrypt ciphertext using recv_key and Header as AAD
        let recv_key = self.recv_key.as_ref().unwrap();
        let header_bytes = &frame[..ENVELOPE_HEADER_SIZE];
        let ciphertext_and_tag = &encrypted_block[NONCE_SIZE..];

        let plaintext = decrypt_aead(recv_key, &nonce, header_bytes, ciphertext_and_tag)?;

        if plaintext.len() < 2 {
            return Err(OnpError::BufferTooShort {
                expected: 2,
                actual: plaintext.len(),
            });
        }

        // 3. Extract physical opcode and resolve to logical opcode
        let physical_opcode = u16::from_le_bytes([plaintext[0], plaintext[1]]);
        let opcode_table = self.opcode_table.as_ref().unwrap();
        let logical_opcode = opcode_table.to_logical(physical_opcode)?;

        let payload_data = plaintext[2..].to_vec();

        Ok(ApplicationPacket {
            opcode: logical_opcode,
            data: payload_data,
        })
    }
}

File: typescript/src/session.ts
/**
 * Session state machine, cryptographic handshake, and anti-replay filter for TypeScript.
 */

import * as crypto from 'crypto';
import {
  ENVELOPE_HEADER_SIZE,
  Flags,
  HANDSHAKE_NONCE_SIZE,
  LogicalOpcode,
  MAC_TAG_SIZE,
  NONCE_SIZE,
} from './constants';
import {
  decryptAead,
  deriveSessionKeys,
  encryptAead,
  KeyExchange,
  SessionKeys,
} from './crypto';
import { constructNonce, extractSequence, EnvelopeHeader } from './framing';
import { OpcodeTable } from './opcode';
import { ApplicationPacket, HandshakeAck, HandshakeSyn } from './packet';

export type SessionRole = 'client' | 'server';
export type SessionState = 'uninitialized' | 'syn_sent' | 'established';

export class AntiReplayWindow {
  private maxSequence: bigint = 0n;
  private bitmap: bigint = 0n;

  /**
   * Non-mutating sequence check against the 64-packet sliding window.
   * MUST be performed before AEAD decryption to filter duplicate or expired frames without mutating state.
   */
  public check(sequence: bigint): boolean {
    if (sequence <= 0n) return false;

    if (sequence > this.maxSequence) {
      return true;
    } else {
      const diff = this.maxSequence - sequence;
      if (diff >= 64n) {
        return false; // Packet fell outside the 64-packet sliding window
      }
      const bit = 1n << diff;
      if ((this.bitmap & bit) !== 0n) {
        return false; // Replay attack detected
      }
      return true;
    }
  }

  /**
   * Commits an authenticated sequence number into the sliding window.
   * MUST only be called AFTER cryptographic MAC tag verification succeeds.
   */
  public commit(sequence: bigint): void {
    if (sequence <= 0n) return;

    if (sequence > this.maxSequence) {
      const diff = sequence - this.maxSequence;
      if (diff < 64n) {
        this.bitmap = (this.bitmap << diff) | 1n;
      } else {
        this.bitmap = 1n;
      }
      this.maxSequence = sequence;
    } else {
      const diff = this.maxSequence - sequence;
      if (diff < 64n) {
        const bit = 1n << diff;
        this.bitmap |= bit;
      }
    }
  }

  /**
   * Legacy convenience helper (atomically checks and commits).
   */
  public checkAndUpdate(sequence: bigint): boolean {
    if (!this.check(sequence)) return false;
    this.commit(sequence);
    return true;
  }
}

export class OnpSession {
  public role: SessionRole;
  public state: SessionState = 'uninitialized';
  private keyExchange: KeyExchange | null = new KeyExchange();
  private localNonce: Buffer = crypto.randomBytes(HANDSHAKE_NONCE_SIZE);
  private peerNonce: Buffer = Buffer.alloc(HANDSHAKE_NONCE_SIZE);
  private sendKey: Buffer | null = null;
  private recvKey: Buffer | null = null;
  private sendSequence: bigint = 0n;
  private replayWindow: AntiReplayWindow = new AntiReplayWindow();
  private opcodeTable: OpcodeTable | null = null;
  private epoch: number = crypto.randomBytes(4).readUInt32LE(0);

  constructor(role: SessionRole) {
    this.role = role;
  }

  public isEstablished(): boolean {
    return this.state === 'established';
  }

  /**
   * CLIENT STEP 1: Initiates handshake by generating HANDSHAKE_SYN frame.
   */
  public createHandshakeSyn(): Buffer {
    if (this.role !== 'client') {
      throw new Error('Server cannot initiate HandshakeSyn');
    }
    if (!this.keyExchange) {
      throw new Error('Key exchange already consumed');
    }

    const syn = new HandshakeSyn(this.keyExchange.publicKey, this.localNonce);
    const payload = syn.encode();
    const header = new EnvelopeHeader(Flags.HANDSHAKE_SYN, payload.length);

    this.state = 'syn_sent';
    return Buffer.concat([Buffer.from(header.encode()), payload]);
  }

  /**
   * SERVER STEP 2: Processes incoming SYN and responds with HANDSHAKE_ACK.
   */
  public processHandshakeSyn(synFrame: Uint8Array): Buffer {
    if (this.role !== 'server') {
      throw new Error('Client cannot process HandshakeSyn as server');
    }

    const header = EnvelopeHeader.decode(synFrame);
    if ((header.flags & Flags.HANDSHAKE_SYN) === 0) {
      throw new Error('Expected HANDSHAKE_SYN flag');
    }

    const synPayload = synFrame.subarray(ENVELOPE_HEADER_SIZE);
    const syn = HandshakeSyn.decode(synPayload);
    this.peerNonce = syn.clientNonce;

    if (!this.keyExchange) {
      throw new Error('Server key exchange already consumed');
    }

    const serverPublicKey = this.keyExchange.publicKey;
    const sharedSecret = this.keyExchange.computeSharedSecret(syn.clientPublicKey);
    this.keyExchange = null; // Erase ephemeral private key immediately

    const keys = deriveSessionKeys(sharedSecret, this.peerNonce, this.localNonce);
    this.sendKey = keys.serverWriteKey;
    this.recvKey = keys.clientWriteKey;
    this.opcodeTable = new OpcodeTable(keys.opcodeSeed);
    this.state = 'established';

    const ack = new HandshakeAck(serverPublicKey, this.localNonce);
    const ackPayload = ack.encode();
    const ackHeader = new EnvelopeHeader(Flags.HANDSHAKE_ACK, ackPayload.length);

    return Buffer.concat([Buffer.from(ackHeader.encode()), ackPayload]);
  }

  /**
   * CLIENT STEP 3: Completes handshake from server's HANDSHAKE_ACK.
   */
  public processHandshakeAck(ackFrame: Uint8Array): void {
    if (this.role !== 'client' || this.state !== 'syn_sent') {
      throw new Error('Client is not in syn_sent state');
    }

    const header = EnvelopeHeader.decode(ackFrame);
    if ((header.flags & Flags.HANDSHAKE_ACK) === 0) {
      throw new Error('Expected HANDSHAKE_ACK flag');
    }

    const ackPayload = ackFrame.subarray(ENVELOPE_HEADER_SIZE);
    const ack = HandshakeAck.decode(ackPayload);
    this.peerNonce = ack.serverNonce;

    if (!this.keyExchange) {
      throw new Error('Client key exchange already consumed');
    }

    const sharedSecret = this.keyExchange.computeSharedSecret(ack.serverPublicKey);
    this.keyExchange = null; // Erase ephemeral private key immediately

    const keys = deriveSessionKeys(sharedSecret, this.localNonce, this.peerNonce);
    this.sendKey = keys.clientWriteKey;
    this.recvKey = keys.serverWriteKey;
    this.opcodeTable = new OpcodeTable(keys.opcodeSeed);
    this.state = 'established';
  }

  /**
   * Encrypts an application command into a complete ONP wire frame.
   */
  public encryptPacket(opcode: LogicalOpcode, payload: Uint8Array): Buffer {
    if (this.state !== 'established' || !this.sendKey || !this.opcodeTable) {
      throw new Error('Session is not established');
    }

    const physicalOpcode = this.opcodeTable.toPhysical(opcode);
    const opcodeBuf = Buffer.alloc(2);
    opcodeBuf.writeUInt16LE(physicalOpcode, 0);

    const plaintext = Buffer.concat([opcodeBuf, Buffer.from(payload)]);

    if (this.sendSequence >= 0xFFFFFFFFFFFFFFFFn) {
      throw new Error('Sequence counter exhausted (nonce reuse protection triggered)');
    }

    this.sendSequence += 1n;
    const nonce = constructNonce(this.epoch, this.sendSequence);

    const totalEncryptedLen = NONCE_SIZE + plaintext.length + MAC_TAG_SIZE;
    const header = new EnvelopeHeader(Flags.ENCRYPTED, totalEncryptedLen);
    const headerBytes = Buffer.from(header.encode());

    const ciphertextAndTag = encryptAead(this.sendKey, nonce, headerBytes, plaintext);

    return Buffer.concat([headerBytes, Buffer.from(nonce), ciphertextAndTag]);
  }

  /**
   * Decrypts an incoming ONP wire frame, checks anti-replay window, and resolves opcode.
   * Enforces strict two-phase validation: non-mutating sequence check prior to AEAD authentication,
   * committing sequence state only after ChaCha20-Poly1305 verification succeeds.
   */
  public decryptPacket(frame: Uint8Array): ApplicationPacket {
    if (this.state !== 'established' || !this.recvKey || !this.opcodeTable) {
      throw new Error('Session is not established');
    }

    const header = EnvelopeHeader.decode(frame);
    if ((header.flags & Flags.ENCRYPTED) === 0) {
      throw new Error('Expected ENCRYPTED flag');
    }

    const expectedTotal = ENVELOPE_HEADER_SIZE + header.payloadLen;
    if (frame.length < expectedTotal) {
      throw new Error(`Buffer too short: expected ${expectedTotal}, got ${frame.length}`);
    }

    const encryptedBlock = frame.subarray(ENVELOPE_HEADER_SIZE, expectedTotal);
    if (encryptedBlock.length < NONCE_SIZE + MAC_TAG_SIZE + 2) {
      throw new Error('Encrypted block too short');
    }

    const nonce = encryptedBlock.subarray(0, NONCE_SIZE);
    const sequence = extractSequence(nonce);

    // 1. Non-mutating sequence check (fails if already seen or behind window)
    if (!this.replayWindow.check(sequence)) {
      throw new Error(`Replay attack detected: sequence ${sequence} rejected by sliding window`);
    }

    const headerBytes = frame.subarray(0, ENVELOPE_HEADER_SIZE);
    const ciphertextAndTag = encryptedBlock.subarray(NONCE_SIZE);

    // 2. Cryptographic AEAD verification
    const plaintext = decryptAead(this.recvKey, nonce, headerBytes, ciphertextAndTag);
    if (plaintext.length < 2) {
      throw new Error('Plaintext too short to contain physical opcode');
    }

    // 3. Commit sequence to sliding window ONLY after authentication succeeds
    this.replayWindow.commit(sequence);

    const physicalOpcode = plaintext.readUInt16LE(0);
    const logicalOpcode = this.opcodeTable.toLogical(physicalOpcode);
    const data = plaintext.subarray(2);

    return {
      opcode: logicalOpcode,
      data,
    };
  }

  /**
   * Zeroizes symmetric keys, nonces, and session state from memory upon teardown.
   */
  public destroy(): void {
    if (this.sendKey) {
      this.sendKey.fill(0);
      this.sendKey = null;
    }
    if (this.recvKey) {
      this.recvKey.fill(0);
      this.recvKey = null;
    }
    this.localNonce.fill(0);
    this.peerNonce.fill(0);
    this.keyExchange = null;
    this.opcodeTable = null;
    this.state = 'uninitialized';
  }
}
